Mathlib/Analysis/NormedSpace/PiTensorProduct/InjectiveSeminorm.lean

```diff
@@ -0,0 +1,467 @@
+we get an induced seminorm on `⨂[𝕜] i, Eᵢ`. We then take the `sup` of these seminorms as
+`F` varies; by `dualSeminorms_bounded`, this family of seminorms is bounded, so the `sup`
+has good properties.
+
+In fact, we cannot take the `sup` over all normed spaces `F` because of set-theoretical issues,
+so we only take spaces `F` in the same universe as `⨂[𝕜] i, Eᵢ`. We then prove in
+`injectiveSeminorm_bound` that this gives the same result, because every multilinear map
+from `E` to `F` factors though a normed vector space in the same universe as
+`⨂[𝕜] i, Eᵢ`.
+
+We then prove the universal property and the functoriality of `⨂[𝕜] i, Eᵢ` as a normed vector
+space.
+
+-- TODO: If all `Eᵢ` are separated and satisfy `SeparatingDual`, then the seminorm on
+`⨂[𝕜] i, Eᵢ` is a norm. This uses the construction of a basis of the `PiTensorProduct`, hence
+depends on PR #11156. It should probably go in a separate file.
+
+## Main definitions
+
+* `PiTensorProduct.toDualContinuousMultilinearMap`: The `𝕜`-linear map from
+`⨂[𝕜] i, Eᵢ` to `ContinuousMultilinearMap 𝕜 E F →L[𝕜] F` sending `x` to the map
+`f ↦ f x`.
+* `PiTensorProduct.injectiveSeminorm`: The injective seminorm on `⨂[𝕜] i, Eᵢ`.
+* `PiTensorProduct.liftEquiv`: The bijection between `ContinuousMultilinearMap 𝕜 E F`
+and `(⨂[𝕜] i, Eᵢ) →L[𝕜] F`, as a continuous linear equivalence.
+* `PiTensorProduct.liftIsometry`: The bijection between `ContinuousMultilinearMap 𝕜 E F`
+and `(⨂[𝕜] i, Eᵢ) →L[𝕜] F`, as an isometric linear equivalence.
+* `PiTensorProduct.tprodL`: The cacnonical continuous multilinear map from `E`
+to `⨂[𝕜] i, Eᵢ`.
+* `PiTensorProduct.mapL`: The continuous linear map from `⨂[𝕜] i, Eᵢ` to `⨂[𝕜] i, E'ᵢ`
+induced by a family of continuous linear maps `Eᵢ →L[𝕜] E'ᵢ`.
+* `PiTensorProduct.mapLMultilinear`: The continuous multilinear map from
+`fun i ↦ (Eᵢ →L[𝕜] E'ᵢ)` to `(⨂[𝕜] i, Eᵢ) →L[𝕜] (⨂[𝕜] i, E'ᵢ)` sending a family
+`f` to `PiTensorProduct.mapL f`.
+
+## Main results
+
+* `PiTensorProduct.injectiveSeminorm_bound`: The main property of the injective seminorm on
+`⨂[𝕜] i, Eᵢ`: for every `x` in `⨂[𝕜] i, Eᵢ` and every continuous multilinear map `f` from `E`
+to a normed space `F`, we have `‖f.lift  x‖ ≤ ‖f‖ * injectiveSeminorm x `.
+* `PiTensorProduct.mapL_opNorm`: If `f` is a family of continuous linear maps
+`fᵢ : Eᵢ →L[𝕜] Fᵢ`, then `‖PiTensorProduct.mapL f‖ ≤ ∏ i, ‖fᵢ‖`.
+* `PiTensorProduct.mapLMultilinear_opNorm` : If `F` is a normed vecteor space, then
+`‖mapLMultilinear 𝕜 E F‖ ≤ 1`.
+
+-/
+
+variable {ι : Type*} [Fintype ι]
+
+variable {𝕜 : Type*} [NontriviallyNormedField 𝕜]
+
+variable {E : ι → Type*} [∀ i, SeminormedAddCommGroup (E i)] [∀ i, NormedSpace 𝕜 (E i)]
+
+variable {F : Type*} [SeminormedAddCommGroup F] [NormedSpace 𝕜 F]
+
+open scoped TensorProduct
+
+open BigOperators
+
+namespace PiTensorProduct
+
+section seminorm
+
+lemma toDualMultilinearMap_bound (x : ⨂[𝕜] i, E i) :
+    ∃ (C : ℝ), 0 ≤ C ∧ ∀ (G : Type*) [SeminormedAddCommGroup G]
```

Comment:
The lemma is a little bit awkward because, as it is written, it depends on some universe (the universe of `G`) and the constant might depend on this universe. Better define the constant inductively depending on `x` (not mentioning `G`, so not depending on any alien universe) and then show that it works, whatever the `G` (and whatever its universe). A bonus is that this explicit constant will give an explicit bound below on the injective seminorm (and maybe `injectiveSeminorm_tprod_le` would be a direct consequence of this).

Reply:
I see what you mean about universes. The "correct" upper bound is actually the projective cross norm on the tensor product, and yes, it does give `injectiveSeminorm_tprod_le`. I'll see if I can define the projective seminorm quickly and use it as an upper bound; it would also have the advantage of being independently interesting, as it would give the definition of the other natural norm on tensor products.